### PR TITLE
Updated Factory and Abstract Factory design patterns, improved code quality

### PR DESCRIPTION
**Pull request by Makas9**

### CODE REVIEW

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/NoOpRunner.cs]: ?

File: Core/NoOpRunner.Core/Shapes/GeometricShape.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/GeometricShape.cs]: I think we are expected to actually make use of the patterns, not just implement empty classes... The classes will have to remain and be used in the final codebase, we cannot just implement random shit and then throw it away.

**Review by Kodnot — CHANGES_REQUESTED** [review]

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/NoOpRunner.cs]: Removed

File: Core/NoOpRunner.Core/Shapes/GeometricShape.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/GeometricShape.cs]: Changed abstract factory design implementation, but problem still exsists. What should I put in these classes?

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/NoOpRunner.cs]: All of these currently render to look like simple platforms. Either add sprites for them, or leave that to @gta-6 and make it so sprites aren't added for these (just render them in different colors)

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/NoOpRunner.cs]: **[Nitpick]** In cases like these, use named arguments (`passable: true)` for better code readability

File: Core/NoOpRunner.Core/Shapes/EntityShape.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/EntityShape.cs:5]: About what to put into these classes. For starters, I was thinking something along the lines of: 
`BaseShape` should have two abstract methods: `bool CanOverlap(BaseShape other)`, `void OnCollision(BaseShape other)`. Child classes override these methods.
For passable shapes (damageCrystals, healthCrystals, passablePlatforms), CanOverlap implementation should return true (maybe with extra movement direction checks for passablePlatforms, not sure about exact impl, can be stubbed for now), OnCollision should triger the appropriate effects (can be stubbed for now).
For impassable shapes, CanOverlap should return false, OnCollision should do damage for saws, explode for rockets, do nothing (or maybe stop movement, exact impl unsure) for platforms.
Then, in theory, we can handle all the shapes and their collisions without explicitly caring about what shapes they are.
But this is only the general direction, you should polish the idea while implementing it, change where need be

File: Core/NoOpRunner.Core/Shapes/ShapeFactories/ShapeFactory.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/ShapeFactories/ShapeFactory.cs:14]: Implementing both abstractFactory and a simple factory for shapes is pretty redundant. On the other hand, AbstractFactory impl already contains a Factory impl. Leave as is for now, but we need to consult the prof about this.

File: Core/NoOpRunner.Core/Shapes/StaticShapes/PassablePlatform.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/StaticShapes/PassablePlatform.cs]: Most of this logic should probably go to StaticShape (see what is/can be common with other static shapes), with only code that's different being in the child classes (as per the other comment)

**Review by Kodnot — CHANGES_REQUESTED** [review]

File: Core/NoOpRunner.Core/Shapes/ShapeFactories/ShapeFactory.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/ShapeFactories/ShapeFactory.cs:14]: Correct, its only here for required pattern implementation

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/NoOpRunner.cs:37]: Yes, I will leave it for @gta-6

File: Core/NoOpRunner.Core/Shapes/EntityShape.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/EntityShape.cs:5]: Implemented empty functions for now.

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/NoOpRunner.cs]: Done

File: Core/NoOpRunner.Core/Shapes/StaticShapes/PassablePlatform.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/StaticShapes/PassablePlatform.cs]: Moved code from child classes

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/NoOpRunner.cs]: I have left this implementation for @gta-6 . I have added code to find which shape is currently being in queue to be painted (GameWindowRenderer.cs)

File: Client/NoOpRunner.Client/Rendering/GameWindowRenderer.cs
Comment (Kodnot) [inline Client/NoOpRunner.Client/Rendering/GameWindowRenderer.cs]: Why use an explicit enumerator? Just `            foreach (var shape in platforms.Shapes)
            {
                var type = shape.GetType();
            }`

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/NoOpRunner.cs]: Same as before, don't pass bool literals like that, use named arguments (`passable: true`). Same for one other place below.

Comment (Kodnot) [inline Core/NoOpRunner.Core/NoOpRunner.cs]: **[Nitpick]** `inpassable` -> `impassable`

File: Core/NoOpRunner.Core/Shapes/BaseShape.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/BaseShape.cs]: These don't belong here, staticShapes neither damage nor heal. Move to EntityShape, and maybe merge into one (maybe `modifyHealth`? Will likely have to refactor anyways when we start implementing the second player)

File: Core/NoOpRunner.Core/Shapes/EntityShape.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/EntityShape.cs]: Replace comments like this with `throw new NotImplementedException();`, here and in other files.

File: Core/NoOpRunner.Core/Shapes/EntityShape.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/EntityShape.cs]: Because the child classes of `EntityShape` do not really have a common implementation for either `CanOverlap` or `OnCollision`, do not implement the methods here (`EntityShape` is abstract anyways, you don't need to). The justification for not inheriting directly from BaseShape can be something like different generation rules in the future for Entities

File: Core/NoOpRunner.Core/Shapes/EntityShapes/DamageCrystal.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/EntityShapes/DamageCrystal.cs]: Use pattern matching. `            if (other is Player p)
            {
                // do stuff with p
            }`. Then, as you cast the shape into type Player, you won't need the `DoDamage()` function on `BaseShape` where it does not belong, you can handle player health changes in player.

File: Core/NoOpRunner.Core/Shapes/EntityShapes/HealthCrystal.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/EntityShapes/HealthCrystal.cs]: Same comment as for `DamageCrystal`

File: Core/NoOpRunner.Core/Shapes/MovingShape.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/MovingShape.cs]: After you merge with changes in dev, `MovingShape` will become abstract, so you can remove these impls

File: Core/NoOpRunner.Core/Shapes/Player.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/Player.cs]: Private fields are camelCase (leave private constants as PascalCase)

File: Core/NoOpRunner.Core/Shapes/Player.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/Player.cs]: As per other comments, leave the functions only on Player (so here) and merge them into one.

File: Core/NoOpRunner.Core/Shapes/PowerUp.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/PowerUp.cs]: **[Q]** Why the change?

File: Core/NoOpRunner.Core/Shapes/PowerUp.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/PowerUp.cs]: You can already implement this part if you want to, just return true if other is Player or Rocket, False otherwise (will need to be updated when we add more Entities)

File: Core/NoOpRunner.Core/Shapes/ShapeFactories/ImpassableShapeFactory.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/ShapeFactories/ImpassableShapeFactory.cs]: Try using more specific exception. This should probably be an `ArgumentException`. Same below and in a few other places.

File: Core/NoOpRunner.Core/Shapes/ShapeFactories/ImpassableShapeFactory.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/ShapeFactories/ImpassableShapeFactory.cs:19]: **[Nitpick]** Add an empty line between methods for readability

File: Core/NoOpRunner.Core/Shapes/StaticShape.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/StaticShape.cs]: Same as with Entity, no need for implementations here, children will override the methods (and you can already implement some of them there, i.e. for a static impassable platform, nothing can overlap it and nothing happens on collision (from the platform's perspective; If i.e. a rocket would hit a platform, the rocket's OnCollision method would be called instead. Or maybe both methods would be called, maybe the logic should go to the objects that it affects, so the logic that explodes the rocket would go to the rocket's OnCollision, the logic that damages the player would go to the player's OnCollision. We'll see later on.)

**Review by Kodnot — CHANGES_REQUESTED** [review]

File: Client/NoOpRunner.Client/Rendering/GameWindowRenderer.cs
Comment (Makas9) [reply Client/NoOpRunner.Client/Rendering/GameWindowRenderer.cs]: Fixed

File: Core/NoOpRunner.Core/NoOpRunner.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/NoOpRunner.cs]: Fixed

Comment (Makas9) [reply Core/NoOpRunner.Core/NoOpRunner.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/BaseShape.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/BaseShape.cs]: Removed this code from BaseShape.cs. Did not move that code to EntityShape for now.

File: Core/NoOpRunner.Core/Shapes/EntityShape.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/EntityShape.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/EntityShape.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/EntityShape.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/EntityShapes/DamageCrystal.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/EntityShapes/DamageCrystal.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/EntityShapes/HealthCrystal.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/EntityShapes/HealthCrystal.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/MovingShape.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/MovingShape.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/Player.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/Player.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/Player.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/Player.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/PowerUp.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/PowerUp.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/PowerUp.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/PowerUp.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/ShapeFactories/ImpassableShapeFactory.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/ShapeFactories/ImpassableShapeFactory.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/ShapeFactories/ImpassableShapeFactory.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/ShapeFactories/ImpassableShapeFactory.cs:19]: Fixed

File: Core/NoOpRunner.Core/Shapes/StaticShape.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/StaticShape.cs]: Fixed. Left empty OnCollision functions in Impassable and Passable platforms.

File: Core/NoOpRunner.Core/Shapes/Player.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/Player.cs]: Could simplify a bit by removing the bool and passing positive/negative values to `healthPoints` instead (could be a one-liner if .net framework had Clamp, sadly that's only on Core)

File: Core/NoOpRunner.Core/Shapes/StaticShapes/ImpassablePlatform.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/StaticShapes/ImpassablePlatform.cs]: **[Nitpick]** Add a `// Do nothing` comment or similar to indicate that this is intentional, and not missing implementation

**Review by Kodnot — CHANGES_REQUESTED** [review]

File: Core/NoOpRunner.Core/Shapes/Player.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/Player.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/StaticShapes/ImpassablePlatform.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/StaticShapes/ImpassablePlatform.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/StaticShape.cs
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/StaticShape.cs]: Currently, the platform generation is f*cked. Revert the ctor changes and set `BottomPosY` and `TopPosY` properly here (and make sure to forward the args properly in child ctors). Right now, those two are not set, so a platform is always generated as a straight line.

**Review by Kodnot — APPROVED** [review]

File: Core/NoOpRunner.Core/Shapes/StaticShape.cs
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/StaticShape.cs]: Fixed

**Pull request merged**
